### PR TITLE
Add new component, DraggablePageFrame, and also update example App.

### PR DESCRIPTION
**Pull request by ivanisidrowu**
## Palumu-lib
- Added new component, `DraggablePageFrame`.
- Added `PageFrameBase` for `DraggablePageFrame` and `ScalablePageFrame`.
- Updated min sdk version from 16 to 19.

## Demo App
- Added new demo activities, `MainActivity` and `DraggableFrameDemoActivity`.
- Changed entry activity from `SimpleVideoListActivity` to `MainActivity`.
- Updated min sdk version from 16 to 19.

## Demo Gif
Click me!

### CODE REVIEW

File: app/src/main/java/tw/invictus/sample/simple/DraggableFrameDemoActivity.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                        handleItemClick()
+                    }
+                }
+            }
+            this.adapter = simpleVideoListAdapter
+        }
+
+        imageView = AppCompatImageView(this)
+        val options = RequestOptions().centerCrop()
+        Glide.with(this).load(R.drawable.sample).apply(options).into(imageView)
```
Comment (HsienDaDa) [inline app/src/main/java/tw/invictus/sample/simple/DraggableFrameDemoActivity.kt]: 單純的resource + center crop
怎不考慮直接 **image.setImageResource(R.drawable.sample)** 搞定呢？

File: palumu-lib/src/main/java/tw/invictus/palumu/DraggablePageFrame.kt [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            clear(view.id)
+            connect(view.id, ConstraintSet.LEFT, ConstraintSet.PARENT_ID, ConstraintSet.LEFT)
+            connect(view.id, ConstraintSet.TOP, ConstraintSet.PARENT_ID, ConstraintSet.TOP)
+            connect(view.id, ConstraintSet.BOTTOM, ConstraintSet.PARENT_ID, ConstraintSet.BOTTOM)
+            connect(view.id, ConstraintSet.RIGHT, ConstraintSet.PARENT_ID, ConstraintSet.RIGHT)
+            constrainWidth(view.id, ConstraintSet.MATCH_CONSTRAINT)
+            constrainHeight(view.id, ConstraintSet.MATCH_CONSTRAINT)
+            applyTo(this@DraggablePageFrame)
+        }
+        invalidate()
```
Comment (HsienDaDa) [inline palumu-lib/src/main/java/tw/invictus/palumu/DraggablePageFrame.kt]: 這邊基本上是view的切換, 非單純同一個view刷新畫面
應該會是 `requestLayout()` 而非 `invalidate()`

File: palumu-lib/src/main/java/tw/invictus/palumu/extension/ViewExtension.kt [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import android.view.View
+
+fun View.isViewHit(target: View, x: Int, y: Int): Boolean {
+    val viewLocation = IntArray(2)
+    target.getLocationOnScreen(viewLocation)
+    val parentLocation = IntArray(2)
+    getLocationOnScreen(parentLocation)
+    val screenX = parentLocation[0] + x
+    val screenY = parentLocation[1] + y
+    return screenX >= viewLocation[0] && screenX < viewLocation[0] + target.width &&
```
Comment (HsienDaDa) [inline palumu-lib/src/main/java/tw/invictus/palumu/extension/ViewExtension.kt]: 後面這邊計算用括號刮起來吧
看是 `(screenX < viewLocation[0] + target.width)` or `screenX < (viewLocation[0] + target.width)`

File: palumu-lib/src/main/java/tw/invictus/palumu/DraggablePageFrame.kt [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    private var contentView: View? = null
+    private var activePointerId = invalidPointer
+    private var lastTouchActionDownXPosition = 0f
+    private var lastTouchActionDownYPosition = 0f
+    private var contentTransition = AutoTransition().apply {
+        duration = animationDuration
+    }
+
+    init {
+        dragHelper = ViewDragHelper.create(this, 1f, DragHelperCallback())
```
Comment (AmyPlace) [inline palumu-lib/src/main/java/tw/invictus/palumu/DraggablePageFrame.kt]: sensitivity可以不用設定，1f是預設值

File: palumu-lib/src/main/java/tw/invictus/palumu/DraggablePageFrame.kt [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+            connect(view.id, ConstraintSet.BOTTOM, ConstraintSet.PARENT_ID, ConstraintSet.BOTTOM)
+            connect(view.id, ConstraintSet.RIGHT, ConstraintSet.PARENT_ID, ConstraintSet.RIGHT)
+            constrainWidth(view.id, ConstraintSet.MATCH_CONSTRAINT)
+            constrainHeight(view.id, ConstraintSet.MATCH_CONSTRAINT)
+            applyTo(this@DraggablePageFrame)
+        }
+        invalidate()
+    }
+
+    fun setContentMargin(margin: Int) {
```
Comment (dev-weiqi) [inline palumu-lib/src/main/java/tw/invictus/palumu/DraggablePageFrame.kt]: [Suggestion]
開給外面設值要告訴外面這個值是dp or px 嗎？
@Dimension(unit = Dimension.DP)
@Dimension(unit = Dimension.PX)

File: palumu-lib/src/main/java/tw/invictus/palumu/DraggablePageFrame.kt [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+        contentView = view
+        view.id = R.id.content
+        startConstraintSet.run {
+            clear(view.id)
+            connect(view.id, ConstraintSet.LEFT, ConstraintSet.PARENT_ID, ConstraintSet.LEFT)
+            connect(view.id, ConstraintSet.TOP, ConstraintSet.PARENT_ID, ConstraintSet.TOP)
+            connect(view.id, ConstraintSet.BOTTOM, ConstraintSet.PARENT_ID, ConstraintSet.BOTTOM)
+            connect(view.id, ConstraintSet.RIGHT, ConstraintSet.PARENT_ID, ConstraintSet.RIGHT)
+            constrainWidth(view.id, ConstraintSet.MATCH_CONSTRAINT)
+            constrainHeight(view.id, ConstraintSet.MATCH_CONSTRAINT)
```
Comment (dev-weiqi) [inline palumu-lib/src/main/java/tw/invictus/palumu/DraggablePageFrame.kt]: [Question] 
#62行 addView 後面已經有給 width/height 了
constrainWidth/constrainHeight 還有必要設定嗎？

File: app/src/main/res/layout/activity_main.xml [hunk 7]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="utf-8"?>
+<android.support.constraint.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    xmlns:app="http://schemas.android.com/apk/res-auto">
+
+    <Button
+        android:id="@+id/scalable_frame_demo"
+        android:layout_width="match_parent"
```
Comment (dev-weiqi) [inline app/src/main/res/layout/activity_main.xml]: left/right有設 constraint 這邊應該可以為 0dp

File: app/src/main/res/layout/activity_main.xml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        android:text="@string/scalable_frame_demo"
+        android:textSize="@dimen/general_text_size"
+        app:layout_constraintLeft_toLeftOf="parent"
+        app:layout_constraintRight_toRightOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        app:layout_constraintBottom_toTopOf="@+id/draggable_frame_demo"/>
+
+    <Button
+        android:id="@+id/draggable_frame_demo"
+        android:layout_width="match_parent"
```
Comment (dev-weiqi) [inline app/src/main/res/layout/activity_main.xml]: left/right有設 constraint 這邊應該可以為 0dp

**Review by dev-weiqi — COMMENTED** [review]
Done

File: palumu-lib/src/main/java/tw/invictus/palumu/DraggablePageFrame.kt (hunk 5 above)
Comment (ivanisidrowu) [reply palumu-lib/src/main/java/tw/invictus/palumu/DraggablePageFrame.kt]: ok

File: palumu-lib/src/main/java/tw/invictus/palumu/DraggablePageFrame.kt (hunk 6 above)
Comment (ivanisidrowu) [reply palumu-lib/src/main/java/tw/invictus/palumu/DraggablePageFrame.kt]: ok

**Review by AmyPlace — APPROVED** [review]

File: palumu-lib/src/main/java/tw/invictus/palumu/DraggablePageFrame.kt [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+            MotionEvent.ACTION_DOWN -> {
+                val index = MotionEventCompat.getActionIndex(ev)
+                activePointerId = MotionEventCompat.getPointerId(ev, index)
+                if (activePointerId == invalidPointer) {
+                    return false
+                }
+            }
+        }
+        val interceptTap = dragHelper.isViewUnder(contentView, ev.x.toInt(), ev.y.toInt())
+        return dragHelper.shouldInterceptTouchEvent(ev) || interceptTap
```
Comment (HsienDaDa) [inline palumu-lib/src/main/java/tw/invictus/palumu/DraggablePageFrame.kt]: 我會把這兩個相反耶
interceptTap是已經被判斷完確定的boolean
不需花任何時間計算了～

**Review by dev-weiqi — APPROVED** [review]

File: palumu-lib/src/main/java/tw/invictus/palumu/DraggablePageFrame.kt [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+        dragHelper.processTouchEvent(ev)
+        content.dispatchTouchEvent(ev)
+        return isDragViewHit
+    }
+
+    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
+        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
+        val view = contentView ?: return
+
+        startConstraintSet.run {
```
Comment (HsienDaDa) [inline palumu-lib/src/main/java/tw/invictus/palumu/DraggablePageFrame.kt:167]: 這段搬過去 onLayout 一起處理吧

File: palumu-lib/src/main/java/tw/invictus/palumu/extension/ViewExtension.kt [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+import android.view.View
+
+fun View.isViewHit(target: View, x: Int, y: Int): Boolean {
+    val viewLocation = IntArray(2)
+    target.getLocationOnScreen(viewLocation)
+    val parentLocation = IntArray(2)
+    getLocationOnScreen(parentLocation)
+    val screenX = parentLocation[0] + x
+    val screenY = parentLocation[1] + y
+    return screenX >= viewLocation[0]
```
Comment (HsienDaDa) [inline palumu-lib/src/main/java/tw/invictus/palumu/extension/ViewExtension.kt]: 調整一下更好理解的寫法

例如：
`screenX >= viewLocation[0] && (screenX < viewLocation[0] + target.width)`
改成
`screenX in viewLocation[0]..viewLocation[0] + target.width` (但這個邊界值需要看一下是否合需求)
or 
`viewLocation[0] >= screenX && screenX < viewLocation[0] + target.width`

**Review by HsienDaDa — APPROVED** [review]

**Pull request merged**